x/mongo/driver/batches.go

```diff
@@ -7,70 +7,114 @@
 package driver
 
 import (
-	"errors"
+	"io"
+	"strconv"
 
 	"go.mongodb.org/mongo-driver/v2/x/bsonx/bsoncore"
+	"go.mongodb.org/mongo-driver/v2/x/mongo/driver/wiremessage"
 )
 
-// ErrDocumentTooLarge occurs when a document that is larger than the maximum size accepted by a
-// server is passed to an insert command.
-var ErrDocumentTooLarge = errors.New("an inserted document is too large")
-
 // Batches contains the necessary information to batch split an operation. This is only used for write
 // operations.
 type Batches struct {
 	Identifier string
 	Documents  []bsoncore.Document
-	Current    []bsoncore.Document
 	Ordered    *bool
-}
-
-// Valid returns true if Batches contains both an identifier and the length of Documents is greater
-// than zero.
-func (b *Batches) Valid() bool { return b != nil && b.Identifier != "" && len(b.Documents) > 0 }
-
-// ClearBatch clears the Current batch. This must be called before AdvanceBatch will advance to the
-// next batch.
-func (b *Batches) ClearBatch() { b.Current = b.Current[:0] }
 
-// AdvanceBatch splits the next batch using maxCount and targetBatchSize. This method will do nothing if
-// the current batch has not been cleared. We do this so that when this is called during execute we
-// can call it without first needing to check if we already have a batch, which makes the code
-// simpler and makes retrying easier.
-// The maxDocSize parameter is used to check that any one document is not too large. If the first document is bigger
-// than targetBatchSize but smaller than maxDocSize, a batch of size 1 containing that document will be created.
-func (b *Batches) AdvanceBatch(maxCount, targetBatchSize, maxDocSize int) error {
-	if len(b.Current) > 0 {
-		return nil
-	}
+	offset int
+}
 
-	if maxCount <= 0 {
-		maxCount = 1
+// AppendBatchSequence appends dst with document sequence of batches as long as the limits of max count, max
+// document size, or total size allows. It returns the number of batches appended, the new appended slice, and
+// any error raised. It returns the origenal input slice if nothing can be appends within the limits.
+func (b *Batches) AppendBatchSequence(dst []byte, maxCount, maxDocSize, _ int) (int, []byte, error) {
```

Comment:
Does `maxMessageSizeBytes` not apply when not using bulk write? For example, when using `collection.Insert` would we batch up to 48mb or 16mb? If I understand the documentation correctly, the batch has a limit of 48mb and individual documents in the batch have a limit of 16mb. Right now it appears we are limiting the entire batch to 16mb.

Reply:
The logic derives from the original `*Batches.AdvanceBatch()`, [which](https://github.com/mongodb/mongo-go-driver/blob/v1/x/mongo/driver/operation.go#L715-L725) takes `maxDocumentSize` as `maxMessageSizeBytes`. A test case covers the logic [here](https://github.com/mongodb/mongo-go-driver/blob/v1/mongo/integration/collection_test.go#L167-L177).

@matthewdale, do you know any history of the logic?

Reply:
I've updated the logic in fc91428.

Reply:
IMO, it doesn't make sense to review incorrect logic since we've already refactored that block. Moreover, fc91428 removed the `maxBsonObjectSize` requirement from the client-level bulk write to simplify the PR.

However, I'm open to reverting and only keeping what is more relevant.

Reply:
IIU the updated solution correctly it is now possible to put a single document over the 16mb limit on the wire message, relying on the server to validate the document size. From what I can tell, this is spec-compliant. But do we want to make round trips that we know are going to fail?

Reply:
Correct, both the client-level bulk write and legacy bulk write rely on the server to check the document size limit in fc91428.

But I agree the validation on the server side is inefficient. Pre-validating the size on the driver side is not a breach of the spec because it is not prohibited, either, especially for performance reasons. So I will commit another change to return driver side errors.

Reply:
FWIW IMO we can defer that work.